### PR TITLE
GH-121832: Assert that the version number of static builtin types is not changed by PyType_Modified.

### PR DESCRIPTION
**Pull request by markshannon**
Update datetime module and test_type_cache.py to not call PyType_Modified.

* Issue: gh-121832

### CODE REVIEW

File: Modules/_datetimemodule.c [hunk 1]
```diff
@@ -7256,49 +7256,51 @@ _datetime_exec(PyObject *module)
...
-    if (PyDict_SetItemString(d, "utc", (PyObject *)&utc_timezone) < 0) {
-        goto error;
-    }
+    if (!reloading) {
+        /* timedelta values */
+        PyObject *d = _PyType_GetDict(&PyDateTime_DeltaType);
+        DATETIME_ADD_MACRO(d, "resolution", new_delta(0, 0, 1, 0));
+        DATETIME_ADD_MACRO(d, "min", new_delta(-MAX_DELTA_DAYS, 0, 0, 0));
+        DATETIME_ADD_MACRO(d, "max",
+                        new_delta(MAX_DELTA_DAYS, 24*3600-1, 1000000-1, 0));
```
Comment (picnixz) [inline Modules/_datetimemodule.c]:
```suggestion
        DATETIME_ADD_MACRO(d, "max",
                           new_delta(MAX_DELTA_DAYS, 24*3600-1, 1000000-1, 0));
```

Nit

File: Modules/_datetimemodule.c [hunk 2]
```diff
@@ -7256,49 +7256,51 @@ _datetime_exec(PyObject *module)
...
+        /* time values */
+        d = _PyType_GetDict(&PyDateTime_TimeType);
+        DATETIME_ADD_MACRO(d, "min", new_time(0, 0, 0, 0, Py_None, 0));
+        DATETIME_ADD_MACRO(d, "max", new_time(23, 59, 59, 999999, Py_None, 0));
+        DATETIME_ADD_MACRO(d, "resolution", new_delta(0, 0, 1, 0));
+
+        /* datetime values */
+        d = _PyType_GetDict(&PyDateTime_DateTimeType);
+        DATETIME_ADD_MACRO(d, "min",
+                        new_datetime(1, 1, 1, 0, 0, 0, 0, Py_None, 0));
```
Comment (picnixz) [inline Modules/_datetimemodule.c]:
```suggestion
        DATETIME_ADD_MACRO(d, "min",
                           new_datetime(1, 1, 1, 0, 0, 0, 0, Py_None, 0));
```

File: Modules/_datetimemodule.c [hunk 3]
```diff
@@ -7256,49 +7256,51 @@ _datetime_exec(PyObject *module)
...
+        DATETIME_ADD_MACRO(d, "max", new_datetime(MAXYEAR, 12, 31, 23, 59, 59,
+                                                999999, Py_None, 0));
```
Comment (picnixz) [inline Modules/_datetimemodule.c]:
```suggestion
        DATETIME_ADD_MACRO(d, "max", new_datetime(MAXYEAR, 12, 31, 23, 59, 59,
                                                  999999, Py_None, 0));
```

**Review by picnixz — COMMENTED** [review]
Just a few nitpicks

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

**Comment by ericsnowcurrently** [comment]
Doesn't this need to be backported to 3.13?

**Pull request merged**